### PR TITLE
[SDPA] New recipe

### PR DESCRIPTION
**Pull request by ViralBShah**

### CODE REVIEW

**Comment by ViralBShah** [comment]
@giordano is there an easy way to get access to `config.log`? I guess I should set up locally, right? Unless you can see something that catches your eye.

**Comment by ViralBShah** [comment]
Actually never mind. I figured it out.

**Comment by ViralBShah** [comment]
The build is downloading mumps and compiling it. I'll fix that. The real issue is the cxx wrapper build:

https://dev.azure.com/JuliaPackaging/Yggdrasil/_build/results?buildId=2815&view=logs&j=5c4d653d-b046-5cfa-092f-8cc5f2272c58&t=f170d2d4-9cee-559b-a6b9-61789383fc3e&l=7992

```
[22:58:52]  ---> cmake -DCMAKE_INSTALL_PREFIX=$prefix -DCMAKE_TOOLCHAIN_FILE=/opt/$target/$target.toolchain -DSDPA_DIR=$prefix -DMUMPS_INCLUDE_DIR="$prefix/include/coin/ThirdParty" -DCMAKE_FIND_ROOT_PATH=$prefix -DJulia_PREFIX=$prefix -DSDPA_LIBRARY="-lsdpa" -DCMAKE_CXX_FLAGS="-march=$processor" -D_GLIBCXX_USE_CXX11_ABI=1 ..
[22:58:52] Re-run cmake no build system arguments
[22:58:52] CMake Error at /usr/share/cmake/Modules/CMakeDetermineSystem.cmake:100 (message):
[22:58:52]   Could not find toolchain file:
[22:58:52]   /opt/aarch64-linux-musl/aarch64-linux-musl.toolchain
[22:58:52] Call Stack (most recent call first):
[22:58:52]   CMakeLists.txt:1 (project)
[22:58:52] 
[22:58:52] 
[22:58:52] CMake Error: CMake was unable to find a build program corresponding to "Unix Makefiles".  CMAKE_MAKE_PROGRAM is not set.  You probably need to select a different build tool.
[22:58:52] CMake Error: CMAKE_C_COMPILER not set, after EnableLanguage
[22:58:52] CMake Error: CMAKE_CXX_COMPILER not set, after EnableLanguage
[22:58:52] -- Configuring incomplete, errors occurred!
[22:58:52]  ---> cmake -DCMAKE_INSTALL_PREFIX=$prefix -DCMAKE_TOOLCHAIN_FILE=/opt/$target/$target.toolchain -DSDPA_DIR=$prefix -DMUMPS_INCLUDE_DIR="$prefix/include/coin/ThirdParty" -DCMAKE_FIND_ROOT_PATH=$prefix -DJulia_PREFIX=$prefix -DSDPA_LIBRARY="-lsdpa" -DCMAKE_CXX_FLAGS="-march=$processor" -D_GLIBCXX_USE_CXX11_ABI=1 ..
[22:58:52] Previous command exited with 1
[22:58:52] Child Process exited, exit code 1
```

How do I fix this?

File: C/Coin-OR/SDPA/build_tarballs.jl [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+if [[ $target == i686-* ]] || [[ $target == arm-* ]]; then
+    export processor=pentium4
+else
+    export processor=x86-64
+fi
+
+cmake -DCMAKE_INSTALL_PREFIX=$prefix -DCMAKE_TOOLCHAIN_FILE=/opt/$target/$target.toolchain -DSDPA_DIR=$prefix -DMUMPS_INCLUDE_DIR="$prefix/include/coin/ThirdParty" \
+-DCMAKE_FIND_ROOT_PATH=$prefix -DJulia_PREFIX=$prefix  -DSDPA_LIBRARY="-lsdpa" -DCMAKE_CXX_FLAGS="-march=$processor" \
+-D_GLIBCXX_USE_CXX11_ABI=1 ..
```
Comment (staticfloat) [inline C/Coin-OR/SDPA/build_tarballs.jl:57]:
```suggestion
cmake -DCMAKE_INSTALL_PREFIX=$prefix \
      -DCMAKE_TOOLCHAIN_FILE=${CMAKE_TARGET_TOOLCHAIN} \
      -DSDPA_DIR=$prefix -DMUMPS_INCLUDE_DIR="$prefix/include/coin/ThirdParty" \
	  -DCMAKE_FIND_ROOT_PATH=${prefix} \
	  -DJulia_PREFIX=${prefix} \
	  -DSDPA_LIBRARY="-lsdpa" \
	  -DCMAKE_CXX_FLAGS="-march=$processor" \
	  -D_GLIBCXX_USE_CXX11_ABI=1
	  ..
```

File: C/Coin-OR/SDPA/build_tarballs.jl [hunk 2]
```diff
@@ -0,0 +1,81 @@
+include("../coin-or-common.jl")
+
+name = "SDPABuilder"
```
Comment (giordano) [inline C/Coin-OR/SDPA/build_tarballs.jl:3]:
```suggestion
name = "SDPA"
```

File: C/Coin-OR/SDPA/build_tarballs.jl [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+cmake -DCMAKE_INSTALL_PREFIX=${prefix} \
+      -DCMAKE_TOOLCHAIN_FILE=${CMAKE_TARGET_TOOLCHAIN} \
+      -DSDPA_DIR=$prefix -DMUMPS_INCLUDE_DIR="$prefix/include/coin/ThirdParty" \
+	  -DCMAKE_FIND_ROOT_PATH=${prefix} \
+	  -DJulia_PREFIX=${prefix} \
+	  -DSDPA_LIBRARY="-lsdpa" \
+	  -DCMAKE_CXX_FLAGS="-march=$processor" \
+	  -D_GLIBCXX_USE_CXX11_ABI=1
+	  ..
```
Comment (giordano) [inline C/Coin-OR/SDPA/build_tarballs.jl:63]:
```suggestion
	  -D_GLIBCXX_USE_CXX11_ABI=1 \
	  ..
```

**Comment by ViralBShah** [comment]
This is now failing because libcxxwrap_julia needs to find Julia. Does it need cross compilation, or just find Julia binaries? Perhaps a question for @barche?

https://dev.azure.com/JuliaPackaging/Yggdrasil/_build/results?buildId=2825&view=logs&j=ad7e3ef1-d50c-5760-c358-6bcad8aee574&t=ca574aa3-17c2-5da7-21b0-f0e395e8b6d3&l=8005

```
 -- Adding path /workspace/destdir to search path
[00:58:20] -- THIS BRANCH
[00:58:20] CMake Error at /workspace/destdir/lib/cmake/JlCxx/FindJulia.cmake:32 (file):
[00:58:20]   file failed to open for reading (No such file or directory):
[00:58:20] 
[00:58:20]     /workspace/srcdir/sdpawrap/Julia_VERSION_INCLUDE-NOTFOUND
[00:58:20] Call Stack (most recent call first):
[00:58:20]   /workspace/destdir/lib/cmake/JlCxx/JlCxxConfig.cmake:3 (find_package)
[00:58:20]   CMakeLists.txt:13 (find_package)
[00:58:20] 
[00:58:20] 
[00:58:20] CMake Error at /workspace/destdir/lib/cmake/JlCxx/FindJulia.cmake:33 (string):
[00:58:20]   string sub-command REGEX, mode MATCH needs at least 5 arguments total to
[00:58:20]   command.
[00:58:20] Call Stack (most recent call first):
[00:58:20]   /workspace/destdir/lib/cmake/JlCxx/JlCxxConfig.cmake:3 (find_package)
[00:58:20]   CMakeLists.txt:13 (find_package)
[00:58:20] 
[00:58:20] 
[00:58:20] -- Julia_VERSION_STRING: 
[00:58:20] CMake Error at /workspace/destdir/lib/cmake/JlCxx/FindJulia.cmake:48 (if):
[00:58:20]   if given arguments:
[00:58:20] 
[00:58:20]     "VERSION_LESS" "0.7.0"
[00:58:20] 
[00:58:20]   Unknown arguments specified
[00:58:20] Call Stack (most recent call first):
[00:58:20]   /workspace/destdir/lib/cmake/JlCxx/JlCxxConfig.cmake:3 (find_package)
[00:58:20]   CMakeLists.txt:13 (find_package)
[00:58:20] 
[00:58:20] 
```

**Comment by barche** [comment]
It just needs to find the Julia binaries, to link against libjulia, i.e. there is no need to run Julia. The elegant way forward would be to have PR #314 completed, but this depends on tarballs for all "official" platforms, preferably named and downloadable from a URL that contains the BB `$target`, so you can download the Julia tarballs using something like `wget https://julialang.org/downloads/julia-1.4.0-$target.tar.gz`

**Comment by ViralBShah** [comment]
Note that #314 is now merged.

**Comment by ViralBShah** [comment]
Rebased to see if this would build now. I don't really understand the errors in here.

File: C/Coin-OR/SDPA/build_tarballs.jl [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+cd build/
+
+if [[ $target == i686-* ]] || [[ $target == arm-* ]]; then
+    export processor=pentium4
+else
+    export processor=x86-64
+fi
+
+cmake -DCMAKE_INSTALL_PREFIX=${prefix} \
+      -DCMAKE_TOOLCHAIN_FILE=${CMAKE_TARGET_TOOLCHAIN} \
```
Comment (barche) [inline C/Coin-OR/SDPA/build_tarballs.jl:56]: Try adding the config options `-DJulia_PREFIX=${prefix} -DCMAKE_FIND_ROOT_PATH=$prefix -DJlCxx_DIR=$prefix/lib/cmake/JlCxx`

File: C/Coin-OR/SDPA/build_tarballs.jl [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+# Dependencies that must be installed before this package can be built
+dependencies = [
+    Dependency("libcxxwrap_julia_jll"),
+    Dependency("OpenBLAS32_jll"),
+    Dependency("CompilerSupportLibraries_jll"),
+]
+
+# Build the tarballs, and possibly a `build.jl` as well.
+build_tarballs(ARGS, name, version, sources, script, platforms, products, dependencies;
+               preferred_gcc_version=gcc_version)
```
Comment (barche) [inline C/Coin-OR/SDPA/build_tarballs.jl:87]: I don't see where `gcc_version` is defined, but it should be at least `v"7"`, and in the log it tries to use GCC 6.

**Comment by ViralBShah** [comment]
Thanks, trying with these.

Does someone know if Mumps is to be included, and if so, which version?

**Comment by ViralBShah** [comment]
Still seems to be having some problem with header files:
```
12:11:06] /workspace/srcdir/sdpawrap/sdpawrap.cpp:1:10: fatal error: 'jlcxx/jlcxx.hpp' file not found
[12:11:06] #include "jlcxx/jlcxx.hpp"
[12:11:06]          ^~~~~~~~~~~~~~~~~
[12:11:07] 1 error generated.
[12:11:07] make[2]: *** [CMakeFiles/sdpawrap.dir/build.make:66: CMakeFiles/sdpawrap.dir/sdpawrap.cpp.o] Error 1
[12:11:07] make[2]: Leaving directory '/workspace/srcdir/sdpawrap/build'
```

**Comment by barche** [comment]
Possibly this is because of the platforms list, copying the list from the libcxxwrap-julia build_tarballs might help.

**Comment by ViralBShah** [comment]
But shouldn't at least some of them pass?

**Comment by barche** [comment]
I seem to remember it's picky about choosing the artifacts because Julia_jll sets the fortran version explicitly, but I'm not 100% sure.

**Pull request closed without merging**